File: backends/fastapi/app/transport/http/dtos.py
"""Wire DTOs. Request bodies are Pydantic models; responses are built as plain dicts by
mapper functions so the JSON matches contract/openapi.yaml exactly (snake_case fields,
RFC3339 timestamps). Keeping response shaping explicit avoids surprises from automatic
serialisation.
"""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel

from app.domain.models import Event, Order, QueueToken, Reservation, Sector
from app.usecase.auth import TokenPair
from app.usecase.events import EventDetail


# --- request models ---

class LoginReq(BaseModel):
    email: str
    password: str


class RefreshReq(BaseModel):
    refresh_token: str


class CreateReservationReq(BaseModel):
    sector_id: str
    quantity: int


class CreateOrderReq(BaseModel):
    reservation_id: str


# --- response mappers ---

def _iso(dt: datetime | None) -> str | None:
    return dt.isoformat() if dt else None


def token_pair(p: TokenPair) -> dict:
    return {
        "access_token": p.access_token,
        "refresh_token": p.refresh_token,
        "token_type": "Bearer",
        "expires_in": p.expires_in,
    }


def event(e: Event) -> dict:
    return {
        "id": e.id,
        "name": e.name,
        "venue": e.venue,
        "starts_at": _iso(e.starts_at),
        "sales_open_at": _iso(e.sales_open_at),
        "status": e.status.value,
    }


def sector(s: Sector) -> dict:
    return {
        "id": s.id,
        "event_id": s.event_id,
        "name": s.name,
        "price_cents": s.price_cents,
        "currency": s.currency,
        "total_inventory": s.total_inventory,
        "available_inventory": s.available_inventory,
    }


def event_detail(d: EventDetail) -> dict:
    return {**event(d.event), "sectors": [sector(s) for s in d.sectors]}


def event_page(events: list[Event], next_cursor: str) -> dict:
    return {"data": [event(e) for e in events], "next_cursor": next_cursor or None}


def queue_token(t: QueueToken) -> dict:
    return {
        "id": t.id,
        "user_id": t.user_id,
        "event_id": t.event_id,
        "position": t.position,
        "status": t.status.value,
        "admitted_at": _iso(t.admitted_at),
    }


def reservation(r: Reservation) -> dict:
    return {
        "id": r.id,
        "user_id": r.user_id,
        "sector_id": r.sector_id,
        "quantity": r.quantity,
        "status": r.status.value,
        "expires_at": _iso(r.expires_at),
    }


def order(o: Order) -> dict:
    return {
        "id": o.id,
        "reservation_id": o.reservation_id,
        "user_id": o.user_id,
        "amount_cents": o.amount_cents,
        "status": o.status.value,
        "created_at": _iso(o.created_at),
    }

File: apps/kmp/sharedUI/src/commonMain/kotlin/com/ticketinglabs/client/data/dto/Dtos.kt
package com.ticketinglabs.client.data.dto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Wire DTOs, one per contract schema. These are generated-in-spirit: they mirror
 * `shared/contract/openapi.yaml` field for field and hold no logic. Enums are kept as raw
 * strings so an unrecognised value is caught by the mapper as a
 * [com.ticketinglabs.client.core.AppError.MalformedResponse] rather than throwing inside
 * the deserializer. Unknown *fields* are ignored by the Json config; unknown *values* in
 * known fields are rejected on mapping. That is the zero-trust posture.
 */

@Serializable
data class ErrorEnvelopeDto(val error: ErrorBodyDto)

@Serializable
data class ErrorBodyDto(
    val code: String,
    val message: String,
    @SerialName("request_id") val requestId: String,
)

@Serializable
data class TokenPairDto(
    @SerialName("access_token") val accessToken: String,
    @SerialName("refresh_token") val refreshToken: String,
    @SerialName("token_type") val tokenType: String,
    @SerialName("expires_in") val expiresIn: Int,
)

@Serializable
data class EventDto(
    val id: String,
    val name: String,
    val venue: String,
    @SerialName("starts_at") val startsAt: String,
    @SerialName("sales_open_at") val salesOpenAt: String,
    val status: String,
)

@Serializable
data class SectorDto(
    val id: String,
    @SerialName("event_id") val eventId: String,
    val name: String,
    @SerialName("price_cents") val priceCents: Int,
    val currency: String,
    @SerialName("total_inventory") val totalInventory: Int,
    @SerialName("available_inventory") val availableInventory: Int,
)

@Serializable
data class EventDetailDto(
    val id: String,
    val name: String,
    val venue: String,
    @SerialName("starts_at") val startsAt: String,
    @SerialName("sales_open_at") val salesOpenAt: String,
    val status: String,
    val sectors: List<SectorDto>,
)

@Serializable
data class EventPageDto(
    val data: List<EventDto>,
    @SerialName("next_cursor") val nextCursor: String? = null,
)

@Serializable
data class QueueTokenDto(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("event_id") val eventId: String,
    val position: Int,
    val status: String,
    @SerialName("admitted_at") val admittedAt: String? = null,
)

@Serializable
data class ReservationDto(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("sector_id") val sectorId: String,
    val quantity: Int,
    val status: String,
    @SerialName("expires_at") val expiresAt: String,
)

@Serializable
data class OrderDto(
    val id: String,
    @SerialName("reservation_id") val reservationId: String,
    @SerialName("user_id") val userId: String,
    @SerialName("amount_cents") val amountCents: Int,
    val status: String,
    @SerialName("created_at") val createdAt: String,
)

// --- Request bodies ---

@Serializable
data class LoginRequestDto(val email: String, val password: String)

@Serializable
data class RefreshRequestDto(@SerialName("refresh_token") val refreshToken: String)

@Serializable
data class CreateReservationDto(
    @SerialName("sector_id") val sectorId: String,
    val quantity: Int,
)

@Serializable
data class CreateOrderDto(@SerialName("reservation_id") val reservationId: String)
